src/routing/kad-dht.md

```diff
@@ -0,0 +1,787 @@
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint
+spec](https://github.com/multiformats/unsigned-varint).
+
+All RPC messages conform to the following protobuf:
+
+```protobuf
+syntax = "proto3";
+
+// Record represents a dht record that contains a value
+// for a key value pair
+message Record {
+    // The key that references this record
+    bytes key = 1;
+
+    // The actual value this record is storing
+    bytes value = 2;
+
+    // Note: These fields were removed from the Record message
+    //
+    // Hash of the authors public key
+    // optional string author = 3;
+    // A PKI signature for the key+value+author
+    // optional bytes signature = 4;
+
+    // Time the record was received, set by receiver
+    // Formatted according to https://datatracker.ietf.org/doc/html/rfc3339
+    string timeReceived = 5;
+};
+
+message Message {
+    enum MessageType {
+        PUT_VALUE = 0;
+        GET_VALUE = 1;
+        ADD_PROVIDER = 2;
+        GET_PROVIDERS = 3;
+        FIND_NODE = 4;
+        PING = 5; // DEPRECATED
+    }
+
+    enum ConnectionType {
+        // sender does not have a connection to peer, and no extra information (default)
+        NOT_CONNECTED = 0;
+
+        // sender has a live connection to peer
+        CONNECTED = 1;
+
+        // sender recently connected to peer
+        CAN_CONNECT = 2;
+
+        // sender recently tried to connect to peer repeatedly but failed to connect
+        // ("try" here is loose, but this should signal "made strong effort, failed")
+        CANNOT_CONNECT = 3;
+    }
+
+    message Peer {
+        // ID of a given peer.
+        bytes id = 1;
+
+        // multiaddrs for a given peer
+        repeated bytes addrs = 2;
+
+        // used to signal the sender's connection capabilities to the peer
+        ConnectionType connection = 3;
+    }
+
+    // defines what type of message it is.
+    MessageType type = 1;
+
+    // defines what coral cluster level this query/response belongs to.
+    // in case we want to implement coral's cluster rings in the future.
+    int32 clusterLevelRaw = 10; // NOT USED
+
+    // Used to specify the key associated with this message.
+    // PUT_VALUE, GET_VALUE, ADD_PROVIDER, GET_PROVIDERS
+    bytes key = 2;
+
+    // Used to return a value
+    // PUT_VALUE, GET_VALUE
+    Record record = 3;
+
+    // Used to return peers closer to a key in a query
+    // GET_VALUE, GET_PROVIDERS, FIND_NODE
+    repeated Peer closerPeers = 8;
+
+    // Used to return Providers
+    // GET_VALUE, ADD_PROVIDER, GET_PROVIDERS
+    repeated Peer providerPeers = 9;
+}
+```
+
+These are the requirements for each `MessageType`:
+
+* `FIND_NODE`: In the request `key` must be set to the binary `PeerId` of the
+node to be found. In the response `closerPeers` is set to the DHT Server's `k`
+closest `Peer`s.
```

Comment:
go-libp2p-kad-dht appears to return [peers that are closer than the requester](https://github.com/libp2p/go-libp2p-kad-dht/blob/010ef38a342f749a6fe380d2b67e651767850f3a/handlers.go#L263C41-L263C48) so we should document this:

```suggestion
closest `Peer`s that are closer to the `key` than the requester's `PeerId`.
```

Reply:
The suggested change doesn't seem right.

When a peer is looking for its closest peers, it will look for its own key. In this case no DHT server will be closest to the requested key than the requester itself. Nevertheless, the requester expects to discover its closest peers.

The response should contain the absolute `k` (number of peers to be returned during lookup) closest peers to the requested key (with exception of self (DHT server), and requester).

IIUC `go-libp2p-kad-dht` doesn't limit the returned peers to those closer than the requester (to the requested key).

Reply:
> IIUC go-libp2p-kad-dht doesn't limit the returned peers to those closer than the requester (to the requested key).

Ah yes, it just passes the requester's peer id in to filter it out and make sure it doesn't tell the requester about itself.